### PR TITLE
Create pl-truth-table element

### PR DESCRIPTION
**Pull request by stac21**
Create a truth table element with support for custom true/false labels, custom rows, inline support of pl elements for variable labels, custom rows and numbers of rows, and user defined expressions. An instance of the element can be found in both the testCourse/elements directory and the top-level elements directory. A training question can be found in the testCourse/questions directory with the name truthTableDemo. The README of this question also has more detailed notes about the element's features and use.

### CODE REVIEW

**Comment by stac21** [comment]
I have reviewed and hereby sign the CLA

**Comment by stac21** [comment]
recheck

File: elements/pl-truth-table/pl-truth-table.css [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+.pl-truth-table-container tr:nth-child(even) {
+	background-color: #f2f2f2;
+}
+
+.pl-truth-table-container td {
+    padding: 1ex;
+    margin: 1em;
+}
```
Comment (SethPoulsen) [inline elements/pl-truth-table/pl-truth-table.css:23]: All files should end with a newline.

File: elements/pl-truth-table/pl-truth-table.css [hunk 2]
```diff
@@ -0,0 +1,23 @@
+.pl-truth-table-container {
```
Comment (SethPoulsen) [inline elements/pl-truth-table/pl-truth-table.css:1]: nice job namespacing your CSS classes!

File: testCourse/elements/pl-truth-table/pl-truth-table.css [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+.pl-truth-table-container tr:nth-child(even) {
+	background-color: #f2f2f2;
+}
+
+.pl-truth-table-container td {
+    padding: 1ex;
+    margin: 1em;
+}
```
Comment (SethPoulsen) [inline testCourse/elements/pl-truth-table/pl-truth-table.css:23]: The element files shouldn't be duplicated in `testCourse`, they should just be in the `elements` folder in the root directory.

File: testCourse/questions/truthTableDemo/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# Training: Truth Table
+
```
Comment (SethPoulsen) [inline testCourse/questions/truthTableDemo/README.md:2]: Documentation should go in `docs/elements.md` and should follow the same style as the documentation for the other elements there.

File: testCourse/questions/truthTableDemo/server.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+def generate(data):
+    variables = [{"name": "A"}, {"name": "B"}, {"name": "C"}, {"name": "D"}]
+
+    rows = []
+
+    rows.append({"C": 0, "D": 1})
+
+    data["params"]["variables"] = variables
+    data["params"]["rows"] = rows
+    data["params"]["expression"] = "A or B"
```
Comment (SethPoulsen) [inline testCourse/questions/truthTableDemo/server.py:10]: Is there a way to use the element without needing to use a `server.py`? It would be nice to have a way to do that, and an example of it.

File: testCourse/questions/truthTableDemo/question.html [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+<pl-truth-table answers-name="something" true-label="on" false-label="off">
+    <pl-variable>C</pl-variable>
+    <pl-variable>D</pl-variable>
+    <pl-variable>E</pl-variable>
+    <pl-answer-column expression="D and E">D and E</pl-answer-column>
+    <pl-row>1 0 1</pl-row>
+    <pl-row>1 1 1</pl-row>
+    <pl-row>0 0 0</pl-row>
+    <pl-row>0 1 0</pl-row>
+</pl-truth-table>
```
Comment (SethPoulsen) [inline testCourse/questions/truthTableDemo/question.html:27]: Questions that are meant to demonstrate the use of an element should be in `exampleCourse/questions/element`. 

The test course is supposed to be just for extra automated tests.

File: elements/pl-truth-table/pl-truth-table.py [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+        score = num_correct / number_answers
+    data["partial_scores"][name] = {"score": score, "weight": weight}
+
+
+def test(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    name = pl.get_string_attrib(element, "answers-name")
+    weight = pl.get_integer_attrib(element, "weight", WEIGHT_DEFAULT)
+
+    _, display_options = data["params"][name]
```
Comment (SethPoulsen) [inline elements/pl-truth-table/pl-truth-table.py:546]: Great to see you've implemented the test function here! Add a line to `tests/exampleCourseQuestions.test.js` so that the test will be run automatically as part of the continuous integration pipeline we have set up.

**Review by SethPoulsen — COMMENTED** [review]
Thanks for your contribution, the element looks really cool!

I've left a bunch of high-level comments, mostly about matching the code style and organization, but haven't had time to read through the code in detail.

File: testCourse/questions/truthTableDemo/server.py (hunk 5 above)
Comment (SethPoulsen) [reply testCourse/questions/truthTableDemo/server.py:10]: ah, ok, I see in the example question, one of the examples uses the `server.py` and one of them doesn't. That's great.

File: testCourse/questions/truthTableDemo/question.html (hunk 6 above)
Comment (SethPoulsen) [reply testCourse/questions/truthTableDemo/question.html:27]: Actually, because this element seems pretty complicated with a bunch of different configuration options, it seems like it could be good to have a question in the example course to show how to use it, and then another one in the test course with a bunch of additional tests covering the different configuration options (for example, see how we have the tests set up for the `pl-order-blocks` element).

Comment (SethPoulsen) [reply testCourse/questions/truthTableDemo/question.html:27]: Or, if you can cover all the edge cases in 3 or 4 examples, just put them all in one file in the example course.

**Comment by SethPoulsen** [comment]
I've approved the CI/build so that when you push more code here it will automatically run the lint/build/test process.

**Comment by SethPoulsen** [comment]
@glherman I know you were working on something similar to this so it would be useful to have your opinion on things like what might work well or not as well for how the HTML syntax is defined.

**Review by nwalters512 — COMMENTED** [review]
This is really exciting! I haven't had a chance to look at the code yet, but @stac21 I was wondering if you'd be able to commit to maintaining this element in the long run. If so, I'd like to add a `CODEOWNERS` file codifying that. Many bits of PL have de facto codeowners, but as we've grown it's become difficult to keep track of them all 😂 

(hi @SethPoulsen, same goes for you and `pl-order-blocks` - I'd love to add you as a codeowner if you're willing!)

**Comment by SethPoulsen** [comment]
> hi @SethPoulsen, same goes for you and `pl-order-blocks` - I'd love to add you as a codeowner if you're willing!

Yes, of course!

**Comment by stac21** [comment]
@nwalters512

**Comment by stac21** [comment]
> This is really exciting! I haven't had a chance to look at the code yet, but @stac21 I was wondering if you'd be able to commit to maintaining this element in the long run. If so, I'd like to add a `CODEOWNERS` file codifying that. Many bits of PL have de facto codeowners, but as we've grown it's become difficult to keep track of them all 😂

For sure!

And please disregard the above closing and opening, I just misclicked the comment button when attempting to post this before.

**Comment by nwalters512** [comment]
@stac21 awesome! If you get your branch up to date with `master`, you should see a `CODEOWNERS` file in the root of the repo. Follow the existing patterns there to add yourself as a codeowner for this new element.

File: elements/pl-truth-table/pl-truth-table.mustache [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    {{#values}}
+        <td>{{html}}</td>
+    {{/values}}
+    {{#answer_columns}}
+        <td>
+            <select name="{{name}}" class="form-control">
+                <option value="True">{{true_label}}</option>
+                <option value="False">{{false_label}}</option>
+            </select>
+        {{#display_score_badge}}
```
Comment (SethPoulsen) [inline elements/pl-truth-table/pl-truth-table.mustache:23]: Can we have the badge on the same line instead of beneath the entry?

File: elements/pl-truth-table/pl-truth-table.py [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+        # A blank is a valid submission from the HTML, but causes a format error.
+        if student_answer == -1:
+            data["format_errors"][
+                expected_html_name
+            ] = "The submitted answer was left blank."
+        elif student_answer is None:
+            data["format_errors"][expected_html_name] = "No answer was submitted."
+
+
+def render(element_html, data):
```
Comment (SethPoulsen) [inline elements/pl-truth-table/pl-truth-table.py:253]: There is a huge amount of code duplication between the `question`, `answer` and `submission` panels. I think a lot of this code can be factored out into a function for code reuse.

**Review by SethPoulsen — COMMENTED** [review]
Ok, I was able to find the time to give a more in-depth review. 

I love the element overall, its going to be great. I think there are some things we can do to really polish up the code before we merge though.

File: elements/pl-truth-table/pl-truth-table.py [hunk 10]
```diff
@@ -0,0 +1,567 @@
...
+    for child in children:
+        if child.tag in ["pl-variable", "pl_variable"]:
+            variable_names.append(pl.inner_html(child))
+
+    return variable_names
+
+
+def get_custom_rows(element):
+    rows = []
+    children = element[:]
```
[the remark this answers is not in the document]
Comment (SethPoulsen) [reply elements/pl-truth-table/pl-truth-table.py:35]: (this comment is refering to the `children = element[:]`)

File: elements/pl-truth-table/pl-truth-table.py [hunk 11]
```diff
@@ -0,0 +1,567 @@
...
+        student_answer = submitted_answers[i]
+
+        if student_answer == correct_answers[i]:
+            num_correct += 1
+
+    score = 0
+    if not partial_credit and num_correct == number_answers:
+        score = 1
+    elif partial_credit:
+        # EDC grading
```
[the remark this answers is not in the document]
Comment (bowmnath) [reply elements/pl-truth-table/pl-truth-table.py:536]: Probably "Every Decision Counts", similar to the pl-checkbox option

File: exampleCourse/questions/element/truthTable/question.html [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+<pl-truth-table answers-name="answer2" true-label="on" false-label="off">
+    <pl-variable>C</pl-variable>
+    <pl-variable>D</pl-variable>
+    <pl-variable>E</pl-variable>
+    <pl-answer-column expression="D and E">D and E</pl-answer-column>
+    <pl-row>1 0 1</pl-row>
+    <pl-row>1 1 1</pl-row>
+    <pl-row>0 0 0</pl-row>
+    <pl-row>0 1 0</pl-row>
+</pl-truth-table>
```
[the remark this answers is not in the document]
Comment (bowmnath) [reply exampleCourse/questions/element/truthTable/question.html:28]: Everything of mine was a Python hack, so I don't have any nice HTML/element structure to compare to.

Is the purpose of those lines simply to limit the rows that show up?  So, only 4 of the 8 rows of that truth table will show up?

If so, maybe it would be easier to have a list attribute such as `show-rows` that took a list of integers and showed the corresponding rows (i.e., the rows with the corresponding binary representations).  In this example, someone could write `show-rows="5,7,0,2"`.

I guess this would be limiting in that you could not have rows with non-binary entries.  I don't recall doing that, but maybe somebody will want to use an `X` or a variable at some point. It doesn't seem likely to me, though.

Just a thought.  Getting back to the current structure, I agree that it looks a little odd with whitespace separating the columns, as opposed to standard HTML, which would have individual tags around every element (`<td>`).  Writing out `<pl-data>` or something for every entry would get very tedious, but I think it would be more standard.

File: exampleCourse/questions/element/truthTable/info.json [hunk 13]
```diff
@@ -0,0 +1,7 @@
+{
+    "uuid": "5150658a-7ec7-4a96-912a-353da738bef0",
+    "title": "Training: Truth Table",
+    "topic": "Training",
+    "tags": ["grantleech", "formative", "easy", "blank", "python", "alpha", "csulb"],
```
[the remark this answers is not in the document]
Comment (stac21) [reply exampleCourse/questions/element/truthTable/info.json:5]: The alpha, blank, and formative tags are all used to indicate what type a question is or what stage of the review process it's in for csulb's questions. I'll go ahead and remove these since they don't serve much of a purpose of that context.

File: exampleCourse/questions/element/truthTable/question.html (hunk 12 above)
[the remark this answers is not in the document]
Comment (stac21) [reply exampleCourse/questions/element/truthTable/question.html:28]: I agree that the whitespace isn't the best, and only came to implement it to avoid requiring something tedious like the `pl-data` tags. I hadn't considered the `show-rows` list attribute, however, and think that that's a great idea!

File: elements/pl-truth-table/pl-truth-table.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (stac21) [reply elements/pl-truth-table/pl-truth-table.py:536]: Exactly, "Every Decision Counts."

File: exampleCourse/questions/element/truthTable/question.html [hunk 14]
```diff
@@ -0,0 +1,33 @@
+<pl-question-panel>
+<p>
+What is the resulting values of the expressions
+<pl-code language="python">y == A or B</pl-code>
+<pl-code language="python">y == C and D</pl-code>
+<pl-code language="python">y == D and E</pl-code>
+<pl-code language="python">y == X or Y</pl-code>
+given the values in the following truth tables?
+</p>
```
[the remark this answers is not in the document]
Comment (stac21) [reply exampleCourse/questions/element/truthTable/question.html:9]: I can see how this can be confusing. After reworking the example question to have descriptions, I think that the question would be better off without this initial prompt and instead should have a short description of the element like in other example questions for elements anyways.

File: elements/pl-truth-table/pl-truth-table.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (stac21) [reply elements/pl-truth-table/pl-truth-table.py:35]: Turns out it isn't necessary, will delete.

File: elements/pl-truth-table/pl-truth-table.py [hunk 15]
```diff
@@ -0,0 +1,567 @@
+import math
```
Comment (eliotwrobson) [inline elements/pl-truth-table/pl-truth-table.py:1]: Not sure if this is being actively worked on still, but for any work going forward, I __strongly__ encourage the use of type hints. It makes your code much easier to read and verify, and also encourages writing cleaner code.

**Comment by stac21** [comment]
@eliotwrobson I got busy with other things so this got put on the back burner, but most of the requested changes are already implemented so I only have a few more things to work on. I have more time now, so I should be able to get it finished by the end of next week. If not, by all means you have free reign to work on it.

**Comment by eliotwrobson** [comment]
@stac21 awesome! If you want, I can commit directly to this PR if you give me write access to this branch on your fork. Otherwise, it would be great if you pushed whatever you have so far so there's no repeated work done. Thanks again!

**Comment by eliotwrobson** [comment]
@stac21 I think I accepted the invite you sent and should have push access now. My semester is over so I should be able to start taking a look. Could you push anything you have responding to the comments so there's no repeated work? Thanks!

File: exampleCourse/questions/element/truthTable/question.html (hunk 12 above)
[the remark this answers is not in the document]
Comment (eliotwrobson) [reply exampleCourse/questions/element/truthTable/question.html:28]: So after looking at the code, I think that the whitespace is not necessary and can be ignored by the code parsing the row tags. It seems like using the `show-rows` attribute as described could make things more difficult for the user, as it requires translation from a bitstring to decimal.

Instead, if the `pl-row` element ignores whitespace, inputs could look like `<pl-row>00101</pl-row>`, with the option to add spaces if the developer finds that easier to read.

[the remark this answers is not in the document]
Comment (SethPoulsen) [reply exampleCourse/questions/element/truthTable/question.html:28]: To me, this is even more of an argument that there should be more explicit tags or separators here, if we are going to have the feature of explicit row specification at all.

Honestly, I think it makes more sense to ship this element _without_ this feature at all, or with the `show-rows` thing that Nathan suggested above. 

Then, after people start using it, if lots of people want this feature where you can explicitly define the rows you want to have and not have, it could be added as a feature.

[the remark this answers is not in the document]
Comment (eliotwrobson) [reply exampleCourse/questions/element/truthTable/question.html:28]: @stac21 Do you find this row selection feature useful in the use cases you have come across? I think it's easy enough to add / remove as-needed, so if people think it's a useful feature then it's worth shipping with the initial version of this element.

[the remark this answers is not in the document]
Comment (SethPoulsen) [reply exampleCourse/questions/element/truthTable/question.html:28]: That's the thing though--once you ship a feature like this to production, its _not_ easy to remove it, because you will break people's existing questions. 

That's why I lean toward don't add the feature until later, when we know people will actually use it.

File: elements/pl-truth-table/pl-truth-table.py [hunk 16]
```diff
@@ -0,0 +1,558 @@
...
+            display_rows.append({
+                "key": str(next(row_counter)),
+                "values": list(map(str, value_assignment))
+            })
+
+            for expr in expressions:
+                #TODO get rid of this eval call!!!
+                val = eval(expr, var_vals)
+                correct_answers.append(str(val))
+
```
Comment (eliotwrobson) [inline elements/pl-truth-table/pl-truth-table.py:124]: For discussion: I was able to substantially simplify the generation logic for variable assignments on each row and the evaluation logic that gets used (should be compatible with a whitelist in sympy).

Before adding more reworked logic here, there seems to be discussion on what row selection options should be available to the question writer as part of this element. I think it seems reasonable to allow custom row selection (as otherwise, this element is functionally capped at 3 variables), but was not sure about what thoughts others have.

**Comment by eliotwrobson** [comment]
@stac21 @SethPoulsen With the last push, I've updated this PR with tests, typechecking compatibility, and some simplifications on the backend. The part of this using the evaluation function can be changed to use a function in the sympy heplers, which is currently being refactored as part of https://github.com/PrairieLearn/PrairieLearn/pull/6641, so this will need to wait for that PR before being merged.

In addition, I simplified the boolean expression evaluation syntax. Since there was some discussion before about what types of row selection features should be included, I figured it would be best to hash that out before going to the trouble of adding those back in. I think some additional test cases can probably also be added, and there might be some room for more table styling options.

**Comment by eliotwrobson** [comment]
Following up on this, the sympy overhaul PR https://github.com/PrairieLearn/PrairieLearn/pull/7020 will introduce a whitelist check function that should work as a whitelist check for this element. Before that gets merged, I'll update the row selection feature based on the discussion here and restart discussion. Then once that PR is merged, it should be easy to update this element to use this function, and having the discussions run concurrently will reduce the time it will take to get this element merged.

**Comment by nwalters512** [comment]
@eliotwrobson @stac21 do either of you want to drive this PR towards being merged?

**Comment by eliotwrobson** [comment]
@nwalters512 I'm happy to push to close if people have use cases for this. I don't think we have any in our course, and this is a pretty niche element overall, so I'm ok with this being closed.

**Comment by nwalters512** [comment]
@eliotwrobson I'll let you make the call! This is definitely a cool element, but yeah, no one seems to be clamoring for it.

**Comment by eliotwrobson** [comment]
@nwalters512 I'll go ahead and close then. As with some other closed PRs, if anyone ever requests something like this it's no trouble to reopen!
